pkg/fsp/header.go

```diff
@@ -0,0 +1,193 @@
+}
+
+// ImageAttribute represents the image attributes.
+type ImageAttribute uint16
+
+func (ia ImageAttribute) String() string {
+	ret := fmt.Sprintf("%#04x GraphicsDisplaySupported", uint16(ia))
+	if uint16(ia) & ^(uint16(1)) != 0 {
+		ret += " (reserved bits are not zeroed)"
+	}
+	return ret
+}
+
+// ComponentAttribute represents the component attribute.
+type ComponentAttribute uint16
+
+func (ca ComponentAttribute) String() string {
+	var attrs []string
+	if uint16(ca)&0x01 == 1 {
+		attrs = append(attrs, "ReleaseBuild")
+	} else {
+		attrs = append(attrs, "DebugBuild")
+	}
+	if uint16(ca)&0x03 == 1 {
+		attrs = append(attrs, "OfficialRelease")
+		attrs = append(attrs, "OfficialRelease")
+	} else {
+		attrs = append(attrs, "TestRelease")
+	}
+	switch uint16(ca) >> 12 {
+	case 1:
+		attrs = append(attrs, "TypeFSP-T")
+	case 2:
+		attrs = append(attrs, "TypeFSP-M")
+	case 3:
+		attrs = append(attrs, "TypeFSP-S")
+	case 8:
+		attrs = append(attrs, "TypeFSP-O")
+	default:
+		attrs = append(attrs, "TypeFSPReserved")
+	}
+	ret := fmt.Sprintf("%#04x %s", uint16(ca), strings.Join(attrs, "|"))
+	// bits 2:12 are reserved
+	if uint16(ca)&0x0ffe != 0 {
+		ret += " (reserved bits are not zeroed)"
+	}
+	return ret
+}
+
+// FromBytes parses an FSP_INFO_HEADER from a byte buffer.
+func FromBytes(b []byte) (*InfoHeaderRev3, error) {
+	if len(b) < CurrentHeaderLength {
+		return nil, fmt.Errorf("short FSP Info Header length %d; want %d", len(b), CurrentHeaderLength)
+	}
+	var f InfoHeaderRev3
+	buf := uio.NewLittleEndianBuffer(b[:CurrentHeaderLength])
+	f.Signature[0] = buf.Read8()
+	f.Signature[1] = buf.Read8()
+	f.Signature[2] = buf.Read8()
+	f.Signature[3] = buf.Read8()
+	f.HeaderLength = buf.Read32()
+	f.Reserved1[0] = buf.Read8()
+	f.Reserved1[1] = buf.Read8()
+	f.SpecVersion = SpecVersion(buf.Read8())
+	f.HeaderRevision = buf.Read8()
+	f.ImageRevision = ImageRevision(buf.Read32())
+	f.ImageID[0] = buf.Read8()
+	f.ImageID[1] = buf.Read8()
+	f.ImageID[2] = buf.Read8()
+	f.ImageID[3] = buf.Read8()
+	f.ImageID[4] = buf.Read8()
+	f.ImageID[5] = buf.Read8()
+	f.ImageID[6] = buf.Read8()
+	f.ImageID[7] = buf.Read8()
+	f.ImageSize = buf.Read32()
+	f.ImageBase = buf.Read32()
+	f.ImageAttribute = ImageAttribute(buf.Read16())
+	f.ComponentAttribute = ComponentAttribute(buf.Read16())
+	f.CfgRegionOffset = buf.Read32()
+	f.CfgRegionSize = buf.Read32()
+	f.Reserved2[0] = buf.Read8()
+	f.Reserved2[1] = buf.Read8()
+	f.Reserved2[2] = buf.Read8()
+	f.Reserved2[3] = buf.Read8()
+	f.TempRAMInitEntryOffset = buf.Read32()
+	f.Reserved3[0] = buf.Read8()
+	f.Reserved3[1] = buf.Read8()
+	f.Reserved3[2] = buf.Read8()
+	f.Reserved3[3] = buf.Read8()
+	f.NotifyPhaseEntryOffset = buf.Read32()
+	f.FSPMemoryInitEntryOffset = buf.Read32()
+	f.TempRAMExitEntryOffset = buf.Read32()
+	f.FSPSiliconInitEntryOffset = buf.Read32()
+	if err := buf.FinError(); err != nil {
+		return nil, err
+	}
+	if !bytes.Equal(f.Signature[:], Signature[:]) {
```

Comment:
does it make sense to move the signature and version checks to the top before the other fields are assigned?

Reply:
indeed, it would, I'll change that.

In the future I plan to also support the previous header revisions, which means first parsing the common header, and then assign to the appropriate implementation of that header revision.